examples/ota-provider-app/ota-provider-common/BdxOtaSender.cpp

```diff
@@ -0,0 +1,144 @@
+ *    limitations under the License.
+ */
+
+#include "BdxOtaSender.h"
+
+#include <core/CHIPError.h>
+#include <messaging/ExchangeContext.h>
+#include <protocols/bdx/BdxTransferSession.h>
+#include <support/BitFlags.h>
+
+#include <fstream>
+
+#include <string.h>
+
+using namespace chip::bdx;
+
+void BdxOtaSender::SetFilepath(const char * path, size_t length)
+{
+    if (path != nullptr)
+    {
+        strcpy(mFilepath, path);
+    }
+    else
+    {
+        memset(mFilepath, 0, kFilepathMaxLength);
+    }
+}
+
+void BdxOtaSender::HandleTransferSessionOutput(TransferSession::OutputEvent & event)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    if (event.EventType != TransferSession::OutputEventType::kNone)
+    {
+        ChipLogDetail(BDX, "OutputEvent type: %d", static_cast<uint16_t>(event.EventType));
+    }
+
+    switch (event.EventType)
+    {
+    case TransferSession::OutputEventType::kNone:
+        break;
+    case TransferSession::OutputEventType::kMsgToSend:
+        VerifyOrReturn(mExchangeCtx != nullptr, ChipLogError(BDX, "%s: mExchangeCtx is null", __FUNCTION__));
+        err = mExchangeCtx->SendMessage(event.msgTypeData.ProtocolId, event.msgTypeData.MessageType, std::move(event.MsgData));
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(BDX, "SendMessage failed: %s", chip::ErrorStr(err));
+        }
+        break;
+    case TransferSession::OutputEventType::kInitReceived: {
+        // TransferSession will automatically reject a transfer if there are no
+        // common supported control modes. It will also default to the smaller
+        // block size.
+        TransferSession::TransferAcceptData acceptData;
+        acceptData.ControlMode  = TransferControlFlags::kReceiverDrive; // OTA must use receiver drive
+        acceptData.MaxBlockSize = mTransfer.GetTransferBlockSize();
+        acceptData.StartOffset  = mTransfer.GetStartOffset();
+        acceptData.Length       = mTransfer.GetTransferLength();
+        VerifyOrReturn(mTransfer.AcceptTransfer(acceptData) == CHIP_NO_ERROR,
+                       ChipLogError(BDX, "%s: %s", __FUNCTION__, chip::ErrorStr(err)));
+        break;
+    }
+    break;
+        break;
+    case TransferSession::OutputEventType::kQueryReceived: {
+        TransferSession::BlockData blockData;
+        uint16_t blockSize   = mTransfer.GetTransferBlockSize();
+        uint16_t bytesToRead = blockSize;
+
+        // TODO: This should be a utility function in TransferSession
+        if (mTransfer.GetTransferLength() > 0 && mNumBytesSent + blockSize > mTransfer.GetTransferLength())
+        {
+            // cast should be safe because of condition above
+            bytesToRead = static_cast<uint16_t>(mTransfer.GetTransferLength() - mNumBytesSent);
+        }
+
+        chip::System::PacketBufferHandle blockBuf = chip::System::PacketBufferHandle::New(bytesToRead);
+
+        std::ifstream otaFile(mFilepath, std::ifstream::in);
```

Comment:
do we need a `.good()` check somewhere here?

Reply:
Sure, that's best